codegen/src/utils/fetch_metadata.rs

```diff
@@ -75,48 +77,122 @@ pub async fn fetch_metadata_hex(
     Ok(hex_data)
 }
 
+async fn fetch_latest_stable(client: impl ClientT) -> Result<Vec<u8>, FetchMetadataError> {
+    // Fetch latest stable metadata of a node via `Metadata_metadata`.
+    let raw: String = client
+        .request("state_call", rpc_params!["Metadata_metadata", "0x"])
+        .await?;
+    let raw_bytes = hex::decode(raw.trim_start_matches("0x"))?;
+    let bytes: frame_metadata::OpaqueMetadata = Decode::decode(&mut &raw_bytes[..])?;
+    Ok(bytes.0)
+}
+
 /// Execute runtime API call and return the specified runtime metadata version.
 async fn fetch_metadata(
     client: impl ClientT,
     version: MetadataVersion,
 ) -> Result<Vec<u8>, FetchMetadataError> {
     const UNSTABLE_METADATA_VERSION: u32 = u32::MAX;
-    const LATEST_STABLE_VERSION: u32 = 14;
 
+    // Note: `Metadata_metadata_versions` may not be present on all nodes.
+    // Once every node supports the new RPC methods, we can simplify the code a bit.
+    let supported: Option<Vec<u32>> = client
+        .request(
+            "state_call",
+            rpc_params!["Metadata_metadata_versions", "0x"],
+        )
+        .await
+        .ok()
+        .map(|raw: String| {
+            let raw_bytes = hex::decode(raw.trim_start_matches("0x"))?;
+            let versions: Vec<u32> = Decode::decode(&mut &raw_bytes[..])?;
+            Ok::<Vec<u32>, FetchMetadataError>(versions)
+        })
+        .transpose()?;
```

Comment:
Does the `?` here lead to us exiting early if this call fails?

Reply:
The code here seems a bit complex after another look, I'll take a note to go back and do a plain-old match to make it easier to follow:

- we initially do a `.ok()` which throws away the `Result<String, Error>` from the client and wraps into a `Option<String>`
- we transpose an Option<Result<..>>, out of which the Result can only be part of the closure: hex decoding or coded::decod  failed

I believe we should be safe here, I'll be a bit more explicit with a simple match and such :D